inst/include/dust/dust.hpp

```diff
@@ -639,6 +808,201 @@ class Dust {
+      std::vector<real_t> y_tmp(ny); // Individual particle state
+      std::vector<real_t> y(np * ny); // Interleaved state of all particles
+      std::vector<uint64_t> rng(np * rng_len); // Interleaved RNG state
+#ifdef _OPENMP
+      #pragma omp parallel for schedule(static) num_threads(_n_threads)
+#endif
+      for (size_t i = 0; i < np; ++i) {
+        // Interleave state
+        _particles[i].state_full(y_tmp.begin());
+        stride_copy(y.data(), y_tmp, i, np);
+
+        // Interleave RNG state
+        dust::rng_state_t<real_t> p_rng = _rng.state(i);
+        size_t rng_offset = i;
+        for (size_t j = 0; j < rng_len; ++j) {
+          rng_offset = stride_copy(rng.data(), p_rng[j], rng_offset, np);
+        }
+      }
+      // H -> D copies
+      _device_data.y.set_array(y);
+      _device_data.rng.set_array(rng);
+      _stale_device = false;
+    }
+  }
+
+  template <typename U = T>
+  typename std::enable_if<dust::has_gpu_support<U>::value, void>::type
+  refresh_host() {
+    if (_stale_host) {
+      const size_t np = n_particles(), ny = n_state_full();
+      const size_t rng_len = dust::rng_state_t<real_t>::size();
+      std::vector<real_t> y_tmp(ny); // Individual particle state
+      std::vector<real_t> y(np * ny); // Interleaved state of all particles
+      std::vector<uint64_t> rngi(np * rng_len); // Interleaved RNG state
+      std::vector<uint64_t> rng(np * rng_len); //  Deinterleaved RNG state
+      // D -> H copies
+      _device_data.y.get_array(y);
+      _device_data.rng.get_array(rngi);
+#ifdef _OPENMP
+      #pragma omp parallel for schedule(static) num_threads(_n_threads)
+#endif
+      for (size_t i = 0; i < np; ++i) {
+        destride_copy(y_tmp.data(), y, i, np);
+        _particles[i].set_state(y_tmp.begin());
+
+        // Destride RNG
+        for (size_t j = 0; j < rng_len; ++j) {
+          rng[i * rng_len + j] = rngi[i + j * np];
+        }
+      }
+      _rng.import_state(rng);
+      _stale_host = false;
+    }
+  }
 };
 
+// TODO: The exact type here for the shared memory will likely change;
+// we'll need to do some translation in dust into native types in
+// order to make the device copy possible. There's nothing really
+// complicated in these in practice.
+template <typename T>
+void run_particles(size_t step_start, size_t step_end, size_t n_particles,
+                   size_t n_state, size_t n_int, size_t n_real, size_t n_pars,
+                   typename T::real_t * state,
+                   std::vector<dust::shared_ptr<T>> shared,
+                   uint64_t * rng_state) {
+  typedef typename T::real_t real_t;
+  const size_t n_particles_each = n_particles / n_pars;
+
+  // Allocation for our internal storage. We won't ever move this to
+  // or from the device so we just pass the size.
+  std::vector<real_t> state_next_v(n_particles * n_state);
+  std::vector<int> internal_int_v(n_particles * n_int);
+  std::vector<real_t> internal_real_v(n_particles * n_real);
```

Comment:
This still needs to be allocated from the host code (even if we are no longer using memset/memcpy), so I think these lines should go back into the dust object, become device_arrays, which are created in initialise. They then need to be passed as pointers to this function. It may make sense to have another struct, like you've done to pack the state + rng?

You can't call CUDA API functions from device or global code sadly.

Reply:
Actually I checked this and `cudaMalloc` and `cudaFree` are both available from device code too, but we still don't want to call these at every kernel entry, so I still think they should be moved back to the dust object.

Reply:
Fixed in ad21b25